### PR TITLE
Use effectiveGasPrice in ante handler for dynamic fee tx

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #814

## Description

Solution:
- use effectiveGasPrice when check minimal-gas-prices, and deduct fee in ante handler
- implement an EthMempoolFeeDecorator

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/types/dynamic_fee_tx.go [hunk 1]
```diff
@@ -256,10 +257,25 @@ func (tx DynamicFeeTx) Validate() error {
...
 func (tx DynamicFeeTx) Cost() *big.Int {
 	return cost(tx.Fee(), tx.GetValue())
 }
+
+// GetEffectiveGasPrice returns the effective gas price
+func (tx *DynamicFeeTx) GetEffectiveGasPrice(baseFee *big.Int) *big.Int {
+	return math.BigMin(new(big.Int).Add(tx.GasTipCap.BigInt(), baseFee), tx.GasFeeCap.BigInt())
+}
+
+// Fee returns gasprice * gaslimit.
```
Comment (thomas-nguy) [inline x/evm/types/dynamic_fee_tx.go:273]: comments needs to be changed accordingly

File: app/ante/eth.go [hunk 2]
```diff
@@ -543,3 +543,55 @@ func (esc EthSetupContextDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simul
...
+		evmDenom := params.EvmDenom
+		if evmtypes.IsLondon(ethCfg, ctx.BlockHeight()) && !feeMktParams.NoBaseFee {
+			baseFee := mfd.feemarketKeeper.GetBaseFee(ctx)
+			feeAmt = msg.GetEffectiveFee(baseFee)
+		} else {
+			feeAmt = msg.GetFee()
+		}
+
+		glDec := sdk.NewDec(int64(msg.GetGas()))
+		requiredFee := ctx.MinGasPrices().AmountOf(evmDenom).Mul(glDec)
```
Comment (thomas-nguy) [inline app/ante/eth.go:590]: this is a local value right? it means all the node won't have the same

Comment (yihuang) [reply app/ante/eth.go:590]: This is a per-validator config value.

File: x/evm/types/dynamic_fee_tx.go (hunk 1 above)
Comment (yihuang) [inline x/evm/types/dynamic_fee_tx.go:273]:
```suggestion
// Fee returns effective_gasprice * gaslimit.
```

File: x/evm/types/dynamic_fee_tx.go [hunk 3]
```diff
@@ -256,10 +257,25 @@ func (tx DynamicFeeTx) Validate() error {
...
+func (tx DynamicFeeTx) EffectiveFee(baseFee *big.Int) *big.Int {
+	return fee(tx.GetEffectiveGasPrice(baseFee), tx.GasLimit)
+}
+
+// Cost returns amount + gasprice * gaslimit.
```
Comment (yihuang) [inline x/evm/types/dynamic_fee_tx.go:278]:
```suggestion
// Cost returns amount + effective_gasprice * gaslimit.
```

File: app/ante/eth.go (hunk 2 above)
Comment (thomas-nguy) [reply app/ante/eth.go:590]: ah okay, maybe add back the comment on top the method
```
	// Ensure that the provided fees meet a minimum threshold for the validator,
	// if this is a CheckTx. This is only for local mempool purposes, and thus
	// is only ran on check tx.
```

File: app/ante/eth.go [hunk 4]
```diff
@@ -543,3 +543,55 @@ func (esc EthSetupContextDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simul
...
+	evmKeeper       EVMKeeper
+}
+
+func NewEthMempoolFeeDecorator(ek EVMKeeper, fmk evmtypes.FeeMarketKeeper) EthMempoolFeeDecorator {
+	return EthMempoolFeeDecorator{
+		feemarketKeeper: fmk,
+		evmKeeper:       ek,
+	}
+}
+
```
Comment (yihuang) [inline app/ante/eth.go:564]:
```suggestion

// AnteHandle ensure that the provided fees meet a minimum threshold for the validator,
// if this is a CheckTx. This is only for local mempool purposes, and thus
// is only ran on check tx.
```

File: app/ante/eth.go [hunk 5]
```diff
@@ -543,3 +543,55 @@ func (esc EthSetupContextDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simul
...
+func (mfd EthMempoolFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	if ctx.IsCheckTx() && !simulate {
```
Comment (yihuang) [inline app/ante/eth.go:566]:
```suggestion
	// Ensure that the provided fees meet a minimum threshold for the validator,
	// if this is a CheckTx. This is only for local mempool purposes, and thus
	// is only ran on check tx.
	if ctx.IsCheckTx() && !simulate {
```

File: app/ante/eth.go (hunk 2 above)
Comment (yihuang) [reply app/ante/eth.go:593]: done

File: rpc/ethereum/namespaces/eth/api.go [hunk 6]
```diff
@@ -843,6 +843,14 @@ func (e *PublicAPI) GetTransactionReceipt(hash common.Hash) (map[string]interfac
 		receipt["contractAddress"] = crypto.CreateAddress(from, txData.GetNonce())
 	}
 
+	if dynamicTx, ok := txData.(*evmtypes.DynamicFeeTx); ok {
+		baseFee, err := e.backend.BaseFee(res.Height)
+		if err != nil {
+			return nil, err
```
Comment (thomas-nguy) [inline rpc/ethereum/namespaces/eth/api.go:849]: not sure if it is good to throw error here,
what if we want to disable basefee but still support eip 1559

Comment (yihuang) [reply rpc/ethereum/namespaces/eth/api.go:849]: > not sure if it is good to throw error here, what if we want to disable basefee but still support eip 1559 transaction type

That's equivalent to enabling basefee, but has a constant base fee amount?

Comment (thomas-nguy) [reply rpc/ethereum/namespaces/eth/api.go:849]: nvm it seems that error is trigger only for internal error

Comment (yihuang) [reply rpc/ethereum/namespaces/eth/api.go:849]: And it don't return error here if base fee is not enabled, it returns a zero and nil.

File: x/evm/types/dynamic_fee_tx.go (hunk 1 above)
Comment (yihuang) [reply x/evm/types/dynamic_fee_tx.go]: thanks, done

File: x/evm/types/dynamic_fee_tx.go [hunk 7]
```diff
@@ -256,10 +257,25 @@ func (tx DynamicFeeTx) Validate() error {
...
 func (tx DynamicFeeTx) Cost() *big.Int {
 	return cost(tx.Fee(), tx.GetValue())
 }
+
+// GetEffectiveGasPrice returns the effective gas price
+func (tx *DynamicFeeTx) GetEffectiveGasPrice(baseFee *big.Int) *big.Int {
+	return math.BigMin(new(big.Int).Add(tx.GasTipCap.BigInt(), baseFee), tx.GasFeeCap.BigInt())
+}
+
+// Fee returns effective_gasprice * gaslimit.
```
Comment (thomas-nguy) [inline x/evm/types/dynamic_fee_tx.go:273]:
```suggestion
// EffectiveFee returns effective_gasprice * gaslimit.
```

File: x/evm/types/dynamic_fee_tx.go [hunk 8]
```diff
@@ -256,10 +257,25 @@ func (tx DynamicFeeTx) Validate() error {
...
+func (tx DynamicFeeTx) EffectiveFee(baseFee *big.Int) *big.Int {
+	return fee(tx.GetEffectiveGasPrice(baseFee), tx.GasLimit)
+}
+
+// Cost returns amount + effective_gasprice * gaslimit.
```
Comment (thomas-nguy) [inline x/evm/types/dynamic_fee_tx.go:278]:
```suggestion
// EffectiveCost returns amount + effective_gasprice * gaslimit.
```

**Review by thomas-nguy — APPROVED** [review]
lgtm

File: app/ante/eth.go [hunk 9]
```diff
@@ -543,3 +543,58 @@ func (esc EthSetupContextDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simul
...
+	return EthMempoolFeeDecorator{
+		feemarketKeeper: fmk,
+		evmKeeper:       ek,
+	}
+}
+
+func (mfd EthMempoolFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	// Ensure that the provided fees meet a minimum threshold for the validator,
+	// if this is a CheckTx. This is only for local mempool purposes, and thus
+	// is only ran on check tx.
```
Comment (fedekunze) [inline app/ante/eth.go:568]: this should be added it as a godoc comment on top of the function

File: app/ante/eth.go [hunk 10]
```diff
@@ -543,3 +543,58 @@ func (esc EthSetupContextDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simul
...
+		if evmtypes.IsLondon(ethCfg, ctx.BlockHeight()) && !feeMktParams.NoBaseFee {
+			baseFee := mfd.feemarketKeeper.GetBaseFee(ctx)
+			feeAmt = msg.GetEffectiveFee(baseFee)
+		} else {
+			feeAmt = msg.GetFee()
+		}
+
+		glDec := sdk.NewDec(int64(msg.GetGas()))
+		requiredFee := ctx.MinGasPrices().AmountOf(evmDenom).Mul(glDec)
+		if !sdk.NewDecFromBigInt(feeAmt).GTE(requiredFee) {
```
Comment (fedekunze) [inline app/ante/eth.go:594]: can you use `LT` instead?

File: x/evm/keeper/utils.go [hunk 11]
```diff
@@ -54,24 +53,19 @@ func (k Keeper) DeductTxCostsFromUserBalance(
...
-	effectiveTip := txData.GetGasPrice()
+	var feeAmt *big.Int
 
 	feeMktParams := k.feeMarketKeeper.GetParams(ctx)
-
 	if london && !feeMktParams.NoBaseFee && txData.TxType() == ethtypes.DynamicFeeTxType {
 		baseFee := k.feeMarketKeeper.GetBaseFee(ctx)
-		gasFeeGap := new(big.Int).Sub(txData.GetGasFeeCap(), baseFee)
-		if gasFeeGap.Sign() == -1 {
+		if txData.GetGasFeeCap().Cmp(baseFee) < 0 {
```
Comment (fedekunze) [inline x/evm/keeper/utils.go:61]: nice 👍

File: x/evm/types/legacy_tx.go [hunk 12]
```diff
@@ -200,3 +200,11 @@ func (tx LegacyTx) Fee() *big.Int {
 func (tx LegacyTx) Cost() *big.Int {
 	return cost(tx.Fee(), tx.GetValue())
 }
+
+func (tx LegacyTx) EffectiveFee(baseFee *big.Int) *big.Int {
```
Comment (fedekunze) [inline x/evm/types/legacy_tx.go:204]: add comments on top of the functions

File: x/evm/types/access_list_tx.go [hunk 13]
```diff
@@ -229,3 +229,11 @@ func (tx AccessListTx) Fee() *big.Int {
 func (tx AccessListTx) Cost() *big.Int {
 	return cost(tx.Fee(), tx.GetValue())
 }
+
+func (tx AccessListTx) EffectiveFee(baseFee *big.Int) *big.Int {
```
Comment (fedekunze) [inline x/evm/types/access_list_tx.go:233]: add comments

**Review by fedekunze — APPROVED** [review]
ACK, minor comments

File: app/ante/eth.go (hunk 9 above)
Comment (yihuang) [reply app/ante/eth.go]: done

File: app/ante/eth.go (hunk 10 above)
Comment (yihuang) [reply app/ante/eth.go]: done

File: x/evm/types/legacy_tx.go (hunk 12 above)
Comment (yihuang) [reply x/evm/types/legacy_tx.go:205]: done

File: x/evm/types/access_list_tx.go (hunk 13 above)
Comment (yihuang) [reply x/evm/types/access_list_tx.go:234]: done

**Pull request merged**